Firebase/InAppMessaging/Flows/FIRIAMMessageClientCache.m

```diff
@@ -126,6 +126,26 @@ - (BOOL)hasTestMessage {
   return self.testMessages.count > 0;
 }
 
+- (nullable FIRIAMMessageDefinition *)nextOnAppLaunchDisplayMsg {
+  // search from the start to end in the list (which implies the display priority) for the
+  // first match (some messages in the cache may not be eligible for the current display
+  // message fetch
+  NSSet<NSString *> *impressionSet =
+      [NSSet setWithArray:[self.bookKeeper getMessageIDsFromImpressions]];
+
+  @synchronized(self) {
+    for (FIRIAMMessageDefinition *next in self.regularMessages) {
+      // message being active and message not impressed yet
+      if ([next messageHasStarted] && ![next messageHasExpired] &&
```

Comment:
can we re-use the logic from the nextOnAppForeground here? (Seems like a code smell that we have to re-implement it fully)

Reply:
More code sharing in 6535db7.